File: frontend/src/api/finmodel.ts
/**
 * FinModel API client — minimal surface needed for the data-entry UI.
 *
 * Only the endpoints the editor actually calls are wrapped here. The backend
 * has 25 endpoints (audit, scenarios, comments, etc.) but those will be added
 * when their respective UI tabs are reconnected.
 */
import { api } from "./client";

const BASE = "/finmodel";

export type RowType = "input" | "subtotal" | "grand" | "check";

export interface TemplateRow {
  code: string;
  section: "BS" | "PL";
  order_idx: number;
  parent_code: string | null;
  row_type: RowType;
  name_ru: string;
  name_uz: string | null;
  name_uz_cyr: string | null;
  name_en: string | null;
  formula: string | null;
  dashboard_category: string | null;
  sign_convention: "positive" | "negative" | null;
  is_indent: number;
  legacy_note: string | null;
}

export interface CellValue {
  row_code: string;
  value: string | null;
  is_calculated: boolean;
  updated_at: string | null;
}

export interface YearLock {
  year: number;
  status: "draft" | "review" | "approved" | "locked";
  locked_at: string | null;
  locked_by: string | null;
  approval_note: string | null;
}

export interface BalanceCheck {
  is_balanced: boolean;
  delta: string;
  asset_total: string;
  liab_total: string;
}

export interface YearData {
  company_id: string;
  year: number;
  lock: YearLock;
  cells: CellValue[];
  balance_check: BalanceCheck;
}

export interface ValidationIssue {
  rule_id: string;
  severity: "error" | "warning" | "info";
  row_code: string | null;
  message_ru: string;
  message_en: string | null;
}

export interface MacroGlobal {
  year: number;
  uz_inflation: string | null;
  us_inflation: string | null;
  uzs_usd_avg_rate: string | null;
  uzs_eur_avg_rate: string | null;
  uzs_rub_avg_rate: string | null;
  uzs_cny_avg_rate: string | null;
  updated_at: string | null;
}

export interface MacroEffective {
  year: number;
  uz_inflation: string | null;
  us_inflation: string | null;
  uzs_usd_avg_rate: string | null;
  uzs_eur_avg_rate: string | null;
  uzs_rub_avg_rate: string | null;
  uzs_cny_avg_rate: string | null;
  source: Record<string, "company" | "global" | "none">;
}

export interface MacroCompanyPayload {
  uz_inflation?: string | null;
  us_inflation?: string | null;
  uzs_usd_avg_rate?: string | null;
  uzs_eur_avg_rate?: string | null;
  uzs_rub_avg_rate?: string | null;
  uzs_cny_avg_rate?: string | null;
}

export interface AuditEntry {
  id: string;
  year: number;
  row_code: string;
  value_before: string | null;
  value_after: string | null;
  actor_id: string | null;
  source: string;
  ts: string;
}

export const finmodelApi = {
  async getTemplate(): Promise<TemplateRow[]> {
    const { data } = await api.get<TemplateRow[]>(`${BASE}/template`);
    return data;
  },
  async listYears(companyId: string): Promise<YearLock[]> {
    const { data } = await api.get<YearLock[]>(`${BASE}/${companyId}`);
    return data;
  },
  async getYear(companyId: string, year: number): Promise<YearData> {
    const { data } = await api.get<YearData>(`${BASE}/${companyId}/${year}`);
    return data;
  },
  async createYear(companyId: string, year: number): Promise<YearLock> {
    const { data } = await api.post<YearLock>(`${BASE}/${companyId}/year/${year}`);
    return data;
  },
  async patchCell(
    companyId: string,
    year: number,
    rowCode: string,
    value: string | null,
  ): Promise<CellValue> {
    const { data } = await api.patch<CellValue>(
      `${BASE}/${companyId}/${year}/cell`,
      { row_code: rowCode, value },
    );
    return data;
  },

  async validate(companyId: string, year: number): Promise<ValidationIssue[]> {
    const { data } = await api.get<ValidationIssue[]>(`${BASE}/${companyId}/${year}/validate`);
    return data;
  },

  async listMacroGlobal(): Promise<MacroGlobal[]> {
    const { data } = await api.get<MacroGlobal[]>(`${BASE}/macro/global`);
    return data;
  },

  async getMacro(companyId: string, year: number): Promise<MacroEffective> {
    const { data } = await api.get<MacroEffective>(`${BASE}/${companyId}/${year}/macro`);
    return data;
  },

  async putMacro(
    companyId: string,
    year: number,
    payload: MacroCompanyPayload,
  ): Promise<MacroEffective> {
    const { data } = await api.put<MacroEffective>(
      `${BASE}/${companyId}/${year}/macro`,
      payload,
    );
    return data;
  },

  async getAudit(
    companyId: string,
    year: number,
    rowCode?: string,
    limit = 100,
  ): Promise<{ items: AuditEntry[]; total: number }> {
    const params: Record<string, string | number> = { limit };
    if (rowCode) params.row_code = rowCode;
    const { data } = await api.get(`${BASE}/${companyId}/${year}/audit`, { params });
    return data;
  },

  async exportCsv(companyId: string, year: number, includeMacro = true): Promise<Blob> {
    const { data } = await api.get(`${BASE}/${companyId}/${year}/export.csv`, {
      params: { include_macro: includeMacro },
      responseType: "blob",
    });
    return data as Blob;
  },

  async deleteYear(companyId: string, year: number): Promise<void> {
    await api.delete(`${BASE}/${companyId}/year/${year}`);
  },

  async lockYear(
    companyId: string,
    year: number,
    status: "review" | "approved" | "locked",
    note?: string,
  ): Promise<YearLock> {
    const { data } = await api.post<YearLock>(
      `${BASE}/${companyId}/year/${year}/lock`,
      { status, approval_note: note ?? null },
    );
    return data;
  },

  async unlockYear(companyId: string, year: number): Promise<YearLock> {
    const { data } = await api.post<YearLock>(`${BASE}/${companyId}/year/${year}/unlock`);
    return data;
  },
};

File: backend/app/schemas/finmodel.py
"""FinModel v2 Pydantic schemas — Phase 1.3."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.schemas._types import MoneyDecimal


# ─── Template ────────────────────────────────────────────────────────
class TemplateRowRead(BaseModel):
    model_config = ConfigDict(from_attributes=True, populate_by_name=True)

    code: str
    section: str
    order_idx: int
    parent_code: Optional[str] = None
    row_type: str
    name_ru: str
    name_uz: Optional[str] = None
    name_uz_cyr: Optional[str] = None
    name_en: Optional[str] = None
    formula: Optional[str] = None
    # Exposed as `dashboard_category` for clarity; persisted in `ifrs_category`
    # column (reusing existing schema until IFRS mapping is implemented).
    dashboard_category: Optional[str] = Field(default=None, validation_alias="ifrs_category")
    sign_convention: Optional[str] = None
    is_indent: int = 0
    legacy_note: Optional[str] = None


# ─── Cell value ──────────────────────────────────────────────────────
class CellValueRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    row_code: str
    value: Optional[MoneyDecimal] = None
    is_calculated: bool = False
    updated_at: Optional[datetime] = None


class CellWrite(BaseModel):
    row_code: str
    value: Optional[MoneyDecimal] = None


class CellBatchWrite(BaseModel):
    cells: list[CellWrite]


# ─── Macro ───────────────────────────────────────────────────────────
class MacroValues(BaseModel):
    uz_inflation: Optional[MoneyDecimal] = None
    us_inflation: Optional[MoneyDecimal] = None
    uzs_usd_avg_rate: Optional[MoneyDecimal] = None
    uzs_eur_avg_rate: Optional[MoneyDecimal] = None
    uzs_rub_avg_rate: Optional[MoneyDecimal] = None
    uzs_cny_avg_rate: Optional[MoneyDecimal] = None


class MacroGlobalRead(MacroValues):
    model_config = ConfigDict(from_attributes=True)
    year: int
    updated_at: Optional[datetime] = None


class MacroCompanyWrite(MacroValues):
    forecast_method: Optional[str] = Field(default="uz_inflation")
    manual_growth_pct: Optional[MoneyDecimal] = None
    dividend_payout_ratio: Optional[MoneyDecimal] = None


class MacroCompanyRead(MacroCompanyWrite):
    model_config = ConfigDict(from_attributes=True)
    company_id: UUID
    year: int
    updated_at: Optional[datetime] = None


class MacroEffective(MacroValues):
    """Resolved macro for a (company, year) — company override falls back to global."""
    year: int
    source: dict[str, str] = Field(default_factory=dict)  # field → 'company' | 'global' | 'none'


# ─── Year lock ───────────────────────────────────────────────────────
class YearLockRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    year: int
    status: str
    locked_at: Optional[datetime] = None
    locked_by: Optional[UUID] = None
    approval_note: Optional[str] = None


class YearLockUpdate(BaseModel):
    status: str
    approval_note: Optional[str] = None


# ─── Year data (full snapshot for one year) ─────────────────────────
class YearDataRead(BaseModel):
    company_id: UUID
    year: int
    lock: YearLockRead
    macro: MacroEffective
    cells: list[CellValueRead]
    # Quick aggregates computed by engine for UI sanity check
    balance_check: dict[str, Any] = Field(default_factory=dict)


# ─── Scenarios ───────────────────────────────────────────────────────
class ScenarioCreate(BaseModel):
    name: str
    description: Optional[str] = None


class ScenarioRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    company_id: UUID
    name: str
    description: Optional[str] = None
    is_active: bool
    created_at: datetime


# ─── Comments ────────────────────────────────────────────────────────
class CommentCreate(BaseModel):
    row_code: str
    comment_text: str
    source_ref: Optional[str] = None


class CommentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    year: int
    row_code: str
    comment_text: str
    source_ref: Optional[str] = None
    author_id: Optional[UUID] = None
    created_at: datetime


# ─── Audit ───────────────────────────────────────────────────────────
class AuditEntry(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: UUID
    year: int
    row_code: str
    value_before: Optional[MoneyDecimal] = None
    value_after: Optional[MoneyDecimal] = None
    actor_id: Optional[UUID] = None
    source: str
    ts: datetime


class AuditList(BaseModel):
    items: list[AuditEntry]
    total: int


# ─── Validation ──────────────────────────────────────────────────────
class ValidationIssue(BaseModel):
    rule_id: str
    severity: str  # error | warning | info
    row_code: Optional[str] = None
    message_ru: str
    message_en: Optional[str] = None


# ─── Forecast ────────────────────────────────────────────────────────
class ForecastRequest(BaseModel):
    base_year: int  # source year (last fact)
    target_years: list[int]  # years to forecast
    method: str = Field(default="uz_inflation")  # uz_inflation | manual | cagr_5y
